### PR TITLE
Add buildconfig proxy admission controller

### PR DESCRIPTION
**Pull request by deads2k**
@bparees @pweil- I needed a break from the rebase, built a quick admission controller that can conditionally set the http proxy env vars and updated the git source to take a pointer to a string and do the same.  If you want it, I can add the integration tests to prove the config works too.

@csrwng you're the natural reviewer if Ben wants to move forward.

### CODE REVIEW

**Comment by bparees** [comment]
@ncdc related to what you're doing w/ forcepull since this needs to cover all the same codepaths.

**Comment by deads2k** [comment]
Running this for `buildconfigs` and `builds` (pretty simple extension) makes sense.  I don't think there's much of a case to run this for other types.

**Comment by bparees** [comment]
so for all who stumble upon this:  this won't work today because not all build object creation flows go through admission control.

File: pkg/build/admission/httpproxy/admission.go [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+		found := false
+		for i := range *envVars {
+			if (*envVars)[i].Name == "HTTP_PROXY" {
+				found = true
+			}
+		}
+		if !found {
+			*envVars = append(*envVars, kapi.EnvVar{Name: "HTTP_PROXY", Value: a.proxyConfig.HTTPProxy})
+		}
```
Comment (csrwng) [inline pkg/build/admission/httpproxy/admission.go]: lines 76-94 are identical to 105-123 except for the name of the env var... you can refactor them to a function

**Comment by csrwng** [comment]
@deads2k @bparees why do we want to apply the proxy to build configs and not builds ? What happens when you change your plugin config? I would suspect we want new builds to pick up the new proxy.

**Comment by bparees** [comment]
@csrwng yes we'll apply it to builds + BCs.  as for what happens when you change the config, that's an excellent question and it would be a good reason to do this as a pod admission controller instead of a build/bc admission controller (intent vs config, as we discussed earlier).  What was the outcome of that discussion for forcepull?

**Comment by csrwng** [comment]
@smarterclayton wants the build generator to use a build store that is wrapped with the admission control chain. In other words, we would mutate the Build.
@ncdc

**Comment by bparees** [comment]
@csrwng but not the buildconfig?  i guess we could do the same here so you'd get new config changes.  except in the case when you create a build from an existing build, since in that case you'd see the old proxy settings as being "set by the user" and not override them.

**Comment by csrwng** [comment]
right not the BuildConfig. And yes, we would not be able to preserve intent on the Build object. An option would be to store the original setting in an annotation?

**Comment by bparees** [comment]
yes that'd be a good option.

**Comment by smarterclayton** [comment]
Do we care about original intent on the build object?  Users
triggering build configs are delegating most of that.

**Comment by bparees** [comment]
we do if they directly created the build object.  which is admittedly less likely.

**Comment by csrwng** [comment]
or if you want to clone the build

**Comment by bparees** [comment]
personally i'm still a fan of doing this in the pod admission controller (as @csrwng suggested).

if it's a build pod, fix up the proxy/env sections appropriately.

preserves original intent, works for any changes to cluster config, and avoids the whole "builds aren't created via the right client" issue.

the only downside is it's not obvious to a developer that their build is using a proxy defined at the cluster level.

**Comment by smarterclayton** [comment]
As a pragmatic solution for short term that's fine.  Since we want to
eventually support project level env, keeping it pod focused is fine.

But don't we still need to properly quota builds on creation time?

**Comment by bparees** [comment]
> But don't we still need to properly quota builds on creation time?

yes, but not seeing how that's related to this?  we already put a quota on the build pod and we're going to put the cgroup on the container we launch.  but that's a separate effort.

**Comment by smarterclayton** [comment]
We have to quota how many build objects you can create?

**Comment by deads2k** [comment]
Ok.  I don't know which bits have to be twiddled on a build pod, so I'll close this.

**Pull request closed without merging**